<!DOCTYPE html>
<html lang="en-us">

  <head>


    <meta charset="UTF-8">
    <title>Flaviviridae-GLUE by giffordlabcvr</title>
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <link rel="stylesheet" type="text/css" href="../stylesheets/normalize.css" media="screen">
    <link href='https://fonts.googleapis.com/css?family=Open+Sans:400,700' rel='stylesheet' type='text/css'>
    <link rel="stylesheet" type="text/css" href="../stylesheets/stylesheet.css" media="screen">
    <link rel="stylesheet" type="text/css" href="../stylesheets/github-light.css" media="screen">

  </head>

  <body>
    <section class="page-header">
      <h1 class="project-name">Flaviviridae-GLUE</h1>
      <h2 class="project-tagline">A GLUE project for Flaviviruses.</h2>
      <a href="../index.html" class="btn">Home</a>
      <a href="https://github.com/giffordlabcvr/Flaviviridae-GLUE/zipball/master" class="btn">Download</a>

      <a href="https://github.com/giffordlabcvr/Flaviviridae-GLUE/tree/master" class="btn">View on GitHub</a>
    </section>

    <section class="main-content">



	   <h3>
		 <a id="Background" class="anchor" href="#Background" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a><strong>Background</strong>
	   </h3>
	   <hr>



	   <br>
	   <h3>
		 <a id="sequenceData" class="anchor" href="#sequenceData" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a><strong>Sequence data</strong>
	   </h3>
	   <hr>

	   <p>
		 The sequence data in this project derive from multiple distinct sources:<br>
	   </p>
	   	  	   <p>	      
		 <b>ncbi-refseqs</b>: Genome-length reference sequences of <a href="https://github.com/giffordlabcvr/Flaviviridae-GLUE/blob/master/tabular/core/cress-refseqs-side-data.tsv">representative flavivirus species</a>.		 
		 These
		 <a href="https://github.com/giffordlabcvr/Flaviviridae-GLUE/tree/master/sources/ncbi-refseqs">XML-formatted files</a>
		  are downloaded directly from <a href="https://www.ncbi.nlm.nih.gov/genbank/">NCBI</a>
		 and are uniquely identified within this project by their <a target="_blank" href="https://www.ncbi.nlm.nih.gov/Sequin/acc.html">GenBank accession numbers</a>. 
	   </p>
	   
	   <p>
		 <b>fasta-curated</b>: A non-redundant set of 
		 <a href="https://github.com/giffordlabcvr/Flaviviridae-GLUE/blob/master/tabular/locus/ecv-locus-data.tsv">flavivirus-derived EVE loci</a>.
		 These
		 <a href="https://github.com/giffordlabcvr/Flaviviridae-GLUE/tree/master/sources/fasta-curated-ecv">FASTA sequences</a>
		 have been curated via systematic screening of whole genome sequence (WGS) assemblies using the 
		 <a target="_blank" href="https://giffordlabcvr.github.io/DIGS-tool/">DIGS tool</a>.
		 Sequences in this source have unique IDs based on arbitrary numbering. 
	   </p>
	      	   
	   <p>
		 <b>fasta-refseqs</b>:
		 
		 <a href="https://github.com/giffordlabcvr/Flaviviridae-GLUE/blob/master/tabular/eve/ecv-refseqs-side-data.tsv">EVE reference sequences</a>
		 - i.e. best-guess estimates of the ancestral sequences of
		 the flaviviruses that gave rise to EVEs.
		 Where possible these are consensus/ancestral sequences derived from alignments
		 <a href="https://github.com/giffordlabcvr/Flaviviridae-GLUE/tree/master/alignments/eve/tip/">included in this project</a>.
		 Sequence IDs used in this source correspond to the names of the unique EVE loci they represent
		 (see <a href="https://retrovirology.biomedcentral.com/articles/10.1186/s12977-018-0442-1" target="_blank">here</a> for details).
	   </p>
 
	   <br>
	   <h3>
	   <a id="sequenceData" class="anchor" href="#sequenceData" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a><strong>Sequence-associated data</strong>
	   </h3>
	   <hr>
	   
	   <p>
         Sequences included in this project are linked to auxiliary data in tabular format, this includes:
		 
		 <ol>
		   <li><a href="https://github.com/giffordlabcvr/Flaviviridae-GLUE/blob/master/tabular/eve/ecv-refseqs-side-data.tsv">Basic taxonomic data</a> for genome-length virus reference sequence in <b>ncbi-refseqs</b>.</li> 
		   <li><a href="https://github.com/giffordlabcvr/Flaviviridae-GLUE/blob/master/tabular/locus/ecv-locus-data.tsv">Locus data</a> for the EVE sequences in <b>fasta-curated</b>.</li> 
		 </ol>
		 
	   </p>
		 
	   <br>
	   <h3>
		 <a id="alignments" class="anchor" href="#alignments" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a><strong>Multiple sequence alignments (MSAs)</strong>
	   </h3>
	   <hr>
	   
	   <p>
		 Several distinct categories of MSA are included in this project, each representing a distinct taxonomic level.
		
		 <ol>
		   <li><b>Tip (i)</b>: Virus species (genome-length alignment)</li> 
		   <li><b>Tip (ii)</b>: <a href="https://github.com/giffordlabcvr/Flaviviridae-GLUE/tree/master/alignments/eve/tip/">EVE lineages</a>.
		   These alignments contain sets of EVE sequences derived from the same ancestral germline colonisation event (i.e. orthologs or duplicates)</li> 
		   <li><b>Internal</b>: <a href="https://github.com/giffordlabcvr/Flaviviridae-GLUE/tree/master/alignments/eve/internal">Virus genera</a></li> 
		   <li><b>Root</b>: <a href="https://github.com/giffordlabcvr/Flaviviridae-GLUE/tree/master/alignments/virus/root">Viruses and EVE reference sequences</a></li> 
		 </ol>
	   </p>






	 <footer class="site-footer">
	   <span class="site-footer-owner"><a href="https://github.com/giffordlabcvr/Flaviviridae-GLUE/">Flaviviridae-GLUE</a> is maintained by <a href="https://github.com/giffordlabcvr">giffordlabcvr</a>.</span>

	   <span class="site-footer-credits">This page was generated by <a href="https://pages.github.com">GitHub Pages</a> using the <a href="https://github.com/jasonlong/cayman-theme">Cayman theme</a> by <a href="https://twitter.com/jasonlong">Jason Long</a>.</span>
	 </footer>

    </section>


  </body>
  
</html>



